wideq/core_v2.py

```diff
@@ -0,0 +1,739 @@
+"""A low-level, general abstraction for the LG SmartThinQ API.
+"""
+import requests
+import base64
+import uuid
+from urllib.parse import urljoin, urlencode, urlparse, parse_qs
+import hashlib
+import hmac
+import logging
+from datetime import datetime
+from typing import Any, Dict, Generator, Optional
+
+from . import as_list, gen_uuid
+from . import core_exceptions as exc
+from . import core
+from .device import DeviceInfo, DEFAULT_TIMEOUT
+
+import os
+import json
+
+# v2
+V2_API_KEY = "VGhpblEyLjAgU0VSVklDRQ=="
+V2_CLIENT_ID = "65260af7e8e6547b51fdccf930097c51eb9885a508d3fddfa9ee6cdec22ae1bd"
+V2_MESSAGE_ID = "wideq"
+V2_SVC_PHASE = "OP"
+V2_APP_LEVEL = "PRD"
+V2_APP_OS = "LINUX"
+V2_APP_TYPE = "NUTS"
+V2_APP_VER = "3.0.1700"
+
+# new
+V2_GATEWAY_URL = "https://route.lgthinq.com:46030/v1/service/application/gateway-uri"
+OAUTH_REDIRECT_URI = "https://kr.m.lgaccount.com/login/iabClose"
+
+# orig
+SECURITY_KEY = "nuts_securitykey"
+SVC_CODE = "SVC202"
+CLIENT_ID = "LGAO221A02"
+OAUTH_SECRET_KEY = "c053c2a6ddeb7ad97cb0eed0dcb31cf8"
+DATE_FORMAT = "%a, %d %b %Y %H:%M:%S +0000"
+
+API2_ERRORS = {
+    "0102": exc.NotLoggedInError,
+    "0106": exc.NotConnectedError,
+}
+
+MIN_TIME_BETWEEN_UPDATE = 25  # seconds
+_LOGGER = logging.getLogger(__name__)
+
+
+def oauth2_signature(message, secret):
+    """Get the base64-encoded SHA-1 HMAC digest of a string, as used in
+    OAauth2 request signatures.
+
+    Both the `secret` and `message` are given as text strings. We use
+    their UTF-8 equivalents.
+    """
+
+    secret_bytes = secret.encode("utf8")
+    hashed = hmac.new(secret_bytes, message.encode("utf8"), hashlib.sha1)
+    digest = hashed.digest()
+    return base64.b64encode(digest)
+
+
+def thinq2_headers(
+    extra_headers={},
+    access_token=None,
+    user_number=None,
+    country=core.DEFAULT_COUNTRY,
+    language=core.DEFAULT_COUNTRY,
+):
+    """Prepare API2 header."""
+
+    headers = {
+        "Accept": "application/json",
+        "Content-type": "application/json;charset=UTF-8",
+        "x-api-key": V2_API_KEY,
+        "x-client-id": V2_CLIENT_ID,
+        "x-country-code": country,
+        "x-language-code": language,
+        "x-message-id": V2_MESSAGE_ID,
+        "x-service-code": SVC_CODE,
+        "x-service-phase": V2_SVC_PHASE,
+        "x-thinq-app-level": V2_APP_LEVEL,
+        "x-thinq-app-os": V2_APP_OS,
+        "x-thinq-app-type": V2_APP_TYPE,
+        "x-thinq-app-ver": V2_APP_VER,
+        "x-thinq-security-key": SECURITY_KEY,
+    }
+
+    if access_token:
+        headers["x-emp-token"] = access_token
+
+    if user_number:
+        headers["x-user-no"] = user_number
+
+    return {**headers, **extra_headers}
+
+
+def thinq2_get(
+    url,
+    access_token=None,
+    user_number=None,
+    headers={},
+    country=core.DEFAULT_COUNTRY,
+    language=core.DEFAULT_COUNTRY,
+):
+    """Make an HTTP request in the format used by the API2 servers."""
+
+    # this code to avoid ssl error 'dh key too small'
```

Comment:
Is this hack necessary for _all_ connections, or is it just in some circumstances that we need this?

Reply:
Just some circumstances, but as before the problem is really understand when should be used

Reply:
Alright, sounds good—but as with other tangential features discussed above, let's move carefully and just do one thing at a time, i.e., leave additional features like this to future pull requests so we can discuss them individually.